File: fetch/request.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"

	"github.com/spf13/viper"
)

var client http.Client = http.Client{}

func MakeRequest(location bool, origin string, query map[string]interface{}) ([]byte, error) {
	var url string
	var err error

	siteInfo := viper.Get("sites").(map[string]interface{})[origin].(map[string]interface{})
	if location {
		url = fmt.Sprintf("https://%s/v3/locations", siteInfo["api"])
	} else {
		url = fmt.Sprintf("https://%s/v2/listings", siteInfo["api"])
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("erro na requisição da página '%s': %v", url, err)
	}

	// Query String
	q := req.URL.Query()
	for key, element := range query {
		q.Add(key, fmt.Sprintf("%v", element))
	}

	req.URL.RawQuery = q.Encode()

	// Headers
	headers := makeHeaders(origin)
	for key, element := range headers {
		req.Header.Add(key, fmt.Sprintf("%v", element))
	}

	// Request
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("erro na requisição da página '%s' %v: %v", url, query, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("erro na requisição da página '%s' %v: status code %v", url, query, http.StatusOK)
	}

	// Response to interface
	bytesData, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("erro no parse da página '%s' %v: %v", url, query, err)
	}

	return bytesData, nil
}

func makeHeaders(origin string) map[string]string {
	return map[string]string{
		"user-agent":       "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36",
		"accept-language":  "pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7",
		"sec-fetch-site":   "same-site",
		"accept":           "application/json",
		"sec-fetch-dest":   "empty",
		"sec-ch-ua-mobile": "?0",
		"sec-fetch-mode":   "cors",
		"origin-ua-mobile": "?0",
		"referer":          fmt.Sprintf("https://www.%s.com.br", origin),
		"origin":           fmt.Sprintf("https://www.%s.com.br", origin),
		"x-domain":         fmt.Sprintf("www.%s.com.br", origin),
	}
}

File: fetch-rust/src/request.rs
use reqwest::header::{HeaderMap, HeaderValue};
use reqwest::{Client, StatusCode};
use std::collections::HashMap;
use std::error::Error;
use url::Url;

pub async fn request(
    url: &str,
    path: &str,
    query: &HashMap<&str, String>,
) -> Result<String, Box<dyn Error>> {
    let client = Client::new();

    let mut headers = HeaderMap::new();
    headers.insert("User-Agent", HeaderValue::from_static("Mozilla/5.0"));
    headers.insert("Accept", HeaderValue::from_static("application/json"));
    headers.insert("x-domain", HeaderValue::from_static(".zapimoveis.com.br"));


    let url = format!("{}{}", url, path);
    let mut url = Url::parse(url.as_str()).expect("Invalid URL");
    for (param, value) in query {
        url.query_pairs_mut().append_pair(param, value);
    }
    let url = url.as_str();
    log::debug!("request url '{}'", url);

    let response = client.get(url).headers(headers).send().await?;

    if response.status() == StatusCode::OK {
        let body = response.text().await?;
        Ok(body)
    } else {
        let url = response.url().to_string();
        let err_msg = format!("Request to '{}' failed with status code: {}", url, response.status());
        Err(err_msg.into())
    }
}
